### PR TITLE
[ValueTracking] Improve KnownBits for signed min-max clamping

### PR DESCRIPTION
**Pull request by adam-bzowski**
A signed min-max clamp is the sequence of smin and smax intrinsics, which constrain a signed value into the range: smin <= value <= smax. The patch improves the calculation of KnownBits for a value subjected to the signed clamping.

### CODE REVIEW

**Comment by adam-bzowski** [comment]
This is a reworking of the solution to the problem investigated in this PR: https://github.com/llvm/llvm-project/pull/119577. Its original goal was reformulated to the problem of type narrowing in the presence of a signed min-max clamp. A signed min-max clamp is the sequence of smin and smax intrinsics, which constrain a signed value into the range: smin <= value <= smax. The patch improves the calculation of KnownBits for a value subjected to the signed clamping.

When calculating KnownBits, whenever smin or smax is encountered, the patch checks if it is a part of the smin-smax clamp. If it is the case, it narrows the KnownBits accordingly. To see why such an improvement was needed, consider a simple clamp of the form

y = smax(smin(x, 0xFF), 0)

for a signed variable x. Without the patch, KnownBits cannot infer anything about smin(x, 100) and thus it only treats y as non-negative (leading bit = 0). With the patch, KnownBits correctly identifies all the leading bits but the 8 lowest ones as vanishing.

The submission essentially consists of 6 additional lines of code (definition of unionWithMinMaxIntrinsicClamp and 2 calls). The remaining change is moving the definition of isSignedMinMaxIntrinsicClamp before its use.

File: llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll
Comment (dtcxzyw) [inline llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll:1]: Please use `llvm/utils/update_test_checks.py`. See also https://llvm.org/docs/InstCombineContributorGuide.html#use-update-test-checks-py.

File: llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll
Comment (dtcxzyw) [inline llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll:7]: Declaration of intrinsics is unnecessary.

File: llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll
Comment (adam-bzowski) [reply llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll:2]: Done. I also added one more vectorial test.

File: llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll
Comment (adam-bzowski) [reply llvm/test/Analysis/ValueTracking/knownbits-trunc-with-min-max-clamp.ll]: Removed.

**Comment by adam-bzowski** [comment]
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

This is fixed as well. Please let me know if I should do something more. Thanks.

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:1122]:
```suggestion
    Known = Known.unionWith(ConstantRange(*CLow, *CHigh + 1).toKnownBits());
```
`*CHigh` should be included by ConstantRange.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: It produces a wrong KnownBits result `00000000000?????` for `smax(smin(X, 32), 0)`.

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: Here is an counterexample: https://alive2.llvm.org/ce/z/ji-_Jq
```
; bin/opt -passes=instcombine test.ll -S
define i16 @test_0a(i16 %x) {
  %1 = tail call i16 @llvm.smin.i16(i16 %x, i16 32)
  %2 = tail call i16 @llvm.smax.i16(i16 %1, i16 0)
  %and = and i16 %2, 31
  ret i16 %and
}
```
Output:
```
define i16 @test_0a(i16 %x) {
  %1 = tail call i16 @llvm.smin.i16(i16 %x, i16 32)
  %2 = tail call i16 @llvm.smax.i16(i16 %1, i16 0)
  ret i16 %2
}
```
`%and` should not be eliminated.

Please add this negative test.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!
Please wait for additional approval from other reviewers :)

**Review by nikic — APPROVED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (topperc) [inline llvm/lib/Analysis/ValueTracking.cpp:1122]: The ConstantRange constructor will assert if CLow and CHigh+1 are the same value.

For example, this incomplete unittest I threw together

```
diff --git a/llvm/unittests/Analysis/ValueTrackingTest.cpp b/llvm/unittests/Analysis/ValueTrackingTest.cpp
index 0145ee70a14c..889f6bd5e78a 100644
--- a/llvm/unittests/Analysis/ValueTrackingTest.cpp
+++ b/llvm/unittests/Analysis/ValueTrackingTest.cpp
@@ -1262,6 +1262,16 @@ TEST_F(ComputeKnownBitsTest, ComputeKnownBits) {
   expectKnownBits(/*zero*/ 4278190085u, /*one*/ 10u);
 }
 
+TEST_F(ComputeKnownBitsTest, ComputeKnownBitsClamp) {
+  parseAssembly(
+      "define i32 @test(i32 %a) {\n"
+      "  %amin = tail call i32 @llvm.smin.i32(i32 %a, i32 2147483647)\n"
+      "  %A = tail call i32 @llvm.smax.i32(i32 %amin, i32 2147483648)\n"
+      "  ret i32 %A\n"
+      "}\n");
+  expectKnownBits(/*zero*/ 0u, /*one*/ 0u);
+}
+
 TEST_F(ComputeKnownBitsTest, ComputeKnownMulBits) {
   parseAssembly(
       "define i32 @test(i32 %a, i32 %b) {\n"
```

Comment (adam-bzowski) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: I don't think that's true. First, *CLow <=  *CHigh as this is checked in isSignedMinMaxIntrinsicClamp. So I understand that you refer to the situation where *CLow is the signed minimum and *CHigh is a signed maximum and it this case

*CLow = *CHigh + 1 = signed minimum.

However, the assert in the contructor of ConstantRange is

assert((Lower != Upper || (Lower.isMaxValue() || Lower.isMinValue())) &&
         "Lower == Upper, but they aren't min or max value!");

So it won't assert, since Lower.isMinValue() is true.

Comment (nikic) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: Nice catch, this should be using `ConstantRange::getNonEmpty()`.

Comment (adam-bzowski) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: This is indeed a bug, I should have added "if" from the very beginning. I made the correction and added further LIT tests to check if everything works fine. Also, it turned out that TruncInstCombine.cpp uses the constraint from the min-max clamp only for some and not all binary operators. I used lshr in the lit tests and this definitely fires the min-max clamp constraint.

Comment (adam-bzowski) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: Here's the fix: https://github.com/llvm/llvm-project/pull/121206. I'm still playing with some tests. Thanks for the comment and sorry for the trouble.

Comment (adam-bzowski) [reply llvm/lib/Analysis/ValueTracking.cpp:1122]: OK, I added the fix and some additional LIT tests in https://github.com/llvm/llvm-project/pull/121206. I added more explanations there.

@nikic If isSignedMinMaxIntrinsicClamp returned true, the range is valid, i.e., *CLow <= *CRight. Thus, *CLow < *CRight + 1, except when *CRight is the max signed value. In such a case *CRight + 1 = min signed value and this is still ok if *CLow is larger than min signed value (the range is a valid interval from *CLow to max signed value).

The problem is when CLow->isMinSignedValue() && CHigh->isMaxSignedValue(). The constructor of ConstantRange has the asserion:

assert((Lower != Upper || (Lower.isMaxValue() || Lower.isMinValue())) &&
         "Lower == Upper, but they aren't min or max value!");

with Lower.isMinValue() and not Lower.isMinSignedValue(). And so it fails.

**Pull request merged**
